## **Resultados Iniciales**

In [ ]:
#@title 1.Importar bibliotecas necesarias
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
from sklearn.metrics import precision_score, recall_score, accuracy_score, f1_score
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix
from sklearn.metrics import mean_squared_error

In [ ]:
#@title 2.Cargar Csv y Visualización
# https://www.kaggle.com/mlg-ulb/creditcardfraud

df = pd.read_csv("creditcard.csv")
df.dropna(axis = 0, inplace=True)
df.drop(['Time'], axis = 1, inplace=True)
df.head()

V1        V2        V3        V4        V5        V6        V7  \
0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9       V10  ...       V21       V22       V23       V24  \
0  0.098698  0.363787  0.090794  ... -0.018307  0.277838 -0.110474  0.066928   
1  0.085102 -0.255425 -0.166974  ... -0.225775 -0.638672  0.101288 -0.339846   
2  0.247676 -1.514654  0.207643  ...  0.247998  0.771679  0.909412 -0.689281   
3  0.377436 -1.387024 -0.054952  ... -0.108300  0.005274 -0.190321 -1.175575   
4 -0.270533  0.817739  0.753074  ... -0.009431  0.798278 -0.137458  0.141267   

        V25       V26       V27       V28  Amount  Class  
0  0.128539 -0.189115  0.133558 -0.021053  149.62      0  
1  0.167170  0.125895 -0.008983  0.014724    2.69      0  
2 -0.327642 -0.139097 -0.055353 -0.059752  378.66      0  
3  0.647376 -0.221929  0.062723  0.061458  123.50      0  
4 -0.206010  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 30 columns]

In [ ]:
df.describe()

V1            V2            V3            V4            V5  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   1.168375e-15  3.416908e-16 -1.379537e-15  2.074095e-15  9.604066e-16   
std    1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00  1.380247e+00   
min   -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00 -1.137433e+02   
25%   -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01 -6.915971e-01   
50%    1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02 -5.433583e-02   
75%    1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01  6.119264e-01   
max    2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01  3.480167e+01   

                 V6            V7            V8            V9           V10  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   1.487313e-15 -5.556467e-16  1.213481e-16 -2.406331e-15  2.239053e-15   
std    1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00  1.088850e+00   
min   -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01 -2.458826e+01   
25%   -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01 -5.354257e-01   
50%   -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02 -9.291738e-02   
75%    3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01  4.539234e-01   
max    7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01  2.374514e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.654067e-16 -3.568593e-16  2.578648e-16  4.473266e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.340915e-16  1.683437e-15 -3.660091e-16 -1.227390e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 30 columns]

In [ ]:
#@title 3.Clases del CSV
df.Class.value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

**Dataset desbalanceado:**
* Clase 0 (No Fraude): 284315 transacciones
* Clase 1 (Fraude): 492 transacciones


In [ ]:
#@title 4.Test y Train
X_train, X_test, y_train, y_test = train_test_split(df.drop("Class", axis=1), df.Class, train_size=.7, random_state=42)


In [ ]:
print("Casos de fraude en test:", sum(y_test))
print("Casos normales en test:", len(y_test) - sum(y_test))


Casos de fraude en test: 136
Casos normales en test: 85307


In [ ]:
#@title 5.Escalar  train y test
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [ ]:
#@title 6.Accuracy: Exactitud de Regresión Logística
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(X_train_scaled, y_train)
y_pred = model.predict(X_test_scaled)
print(accuracy_score(y_test, y_pred))

0.9992626663389628


Aunque el modelo tiene una exactitud (accuracy) muy alta de 99.92%, **esto no significa que sea un buen modelo para detectar fraudes**.

In [ ]:
#@title 7.Matriz de Confusión:
#Podemos evaluar que no es tan buena
conf = confusion_matrix(y_test, y_pred)
conf

array([[85295,    12],
       [   51,    85]])

* 85295 TN (True Negatives) → **85295 transacciones normales detectadas** correctamente.
* 12 FP (False Positives) → **12 transacciones normales** clasificadas como fraude.
* 51 FN (False Negatives) → **51 fraudes no detectados** (el modelo falló en ellos).
* 85 TP (True Positives) → **85 fraudes detectados** correctamente.


In [ ]:
#@title 8.Métricas de Evaluación
print("Accuracy: ", accuracy_score(y_test, y_pred))
print("Recall: ", recall_score(y_test, y_pred))
print("precision_score: ", precision_score(y_test, y_pred))
print("Especificidad: ", recall_score(y_test, y_pred, pos_label=0)) # Negativos predichos correctamente
print("F1-score: ", f1_score(y_test, y_pred))
print("MSE: ", mean_squared_error(y_test, y_pred))


Accuracy:  0.9992626663389628
Recall:  0.625
precision_score:  0.8762886597938144
Especificidad:  0.999859331590608
F1-score:  0.7296137339055794
MSE:  0.0007373336610371826


### **Evaluación personal de las métricas**
**ASPECTOS POSITIVOS:**

* **Specificity (Especificidad) = 0.99986 (~99.99%)**: Detecta perfectamente los casos normales.
* **F1-score = 0.729 (~72.96%)**: Un balance entre precisión y recall, el valor no es malo, pero tampoco ideal.

**ASPECTOS NEGATIVOS:**
* **Accuracy (Exactitud) = 0.99926 (~99.93%)** Métrica engañosa porque el dataset está desbalanceado, la **alta exactitud no significa que detecte bien los fraudes**.
* **Recall (Sensibilidad) = 0.625 (~62.5%):** De cada 100 fraudes reales, detecta 62 y deja escapar 38.**Queremos maximizar el Recall.**
* **Precision (Precisión) = 0.876 (~87.6%)** : Cuando el modelo predice "Fraude", acierta el 87.6% de las veces.

**SIGUIENTE PASO:** Usar técnicas de detección de anomalías para **mejorar el recall.**

De las transacciones fraudulentas, aunque pocás, ha acertado poco menos de la mitad

# **Detectar Anomalías para Mejorar la Predicción**

---


### **¿Qué estamos intentando hacer?**
Queremos detectar transacciones fraudulentas en un conjunto de datos de tarjetas de crédito. Sin embargo, hay un problema:

* El fraude es raro (muy pocos casos en comparación con las transacciones normales).
* Si entrenamos un modelo normal, detectará muy bien las transacciones normales, pero fallará en detectar fraudes.
* Necesitamos un método que encuentre anomalías (cosas que se salen de lo común), porque los fraudes son anomalías.
* Por eso usamos Determinante de Covarianza Mínima (MCD) y otros métodos de detección de anomalías.

## **1. Pregunta**
Utiliza el **Determinante de Covarianza Mínima** para detectar los outliers y hacer predicciones con la misma. Calcula todas las métricas anteriores y muestra la matriz de confusión

---



**¿Qué es el Determinante de Covarianza Mínima (MCD)?**

Es un método que busca los datos que mejor representan el "comportamiento normal" y **descarta los valores extremos (outliers)**, que podrían ser fraudes.

Analiza todas las transacciones y busca las que parecen "normales".
Descarta las transacciones más raras, porque probablemente sean fraudes.
Entrena un modelo con solo los datos normales.
Luego, usa ese modelo para hacer predicciones en los datos de prueba.

In [ ]:
#@title 1.Determinante de Covarianza:
from sklearn.covariance import EllipticEnvelope

# Lista para almacenar las métricas de cada iteración
results = []

# Iterar sobre diferentes niveles de contaminación
for c in np.arange(0.01, 0.3, 0.01):
    # Aplicar el modelo EllipticEnvelope con la contaminación actual
    mcd = EllipticEnvelope(contamination=c, random_state=42)
    mcd.fit(X_train_scaled)

    # Detectar outliers y filtrar los datos
    y_pred_elli = mcd.predict(X_train_scaled)
    mask = y_pred_elli != -1  # Filtrar solo los datos no considerados outliers
    X_train_elli, y_train_elli = X_train_scaled[mask], y_train[mask]

    # Entrenar el modelo de Regresión Logística
    model_mcd = LogisticRegression()
    model_mcd.fit(X_train_elli, y_train_elli)

    # Predecir en el conjunto de prueba
    y_pred_mcd = model_mcd.predict(X_test_scaled)

    # Calcular métricas
    metrics = {
        "Contaminación": c,
        "Accuracy": accuracy_score(y_test, y_pred_mcd),
        "Recall": recall_score(y_test, y_pred_mcd),
        "Precision": precision_score(y_test, y_pred_mcd),
        "Specificity": recall_score(y_test, y_pred_mcd, pos_label=0),  # Negativos predichos correctamente
        "F1-score": f1_score(y_test, y_pred_mcd)
    }

    results.append(metrics)

Streaming output truncated to the last 5000 lines.
/usr/local/lib/python3.11/dist-packages/sklearn/covariance/_robust_covariance.py:185: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-191.218180695156036 > -298.257078914440854). You may want to try with a higher value of support_fraction (current value: 0.501).
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/covariance/_robust_covariance.py:185: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-253.553537586707506 > -254.462736667838271). You may want to try with a higher value of support_fraction (current value: 0.501).
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/covariance/_robust_covariance.py:185: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-248.238403078166982 > -248.800545362113724). You may want to try with a higher value of support_fraction 

In [ ]:
#@title 2.Mostrar Dataframe con  Métricas para cada tipo de Contaminación
from IPython.display import display

# Convertir a DataFrame
metrics_df = pd.DataFrame(results)

# Mostrar la tabla en el notebook
display(metrics_df)

Contaminación  Accuracy    Recall  Precision  Specificity  F1-score
0            0.01  0.999380  0.808824   0.802920     0.999683  0.805861
1            0.02  0.998818  0.816176   0.593583     0.999109  0.687307
2            0.03  0.997507  0.808824   0.370370     0.997808  0.508083
3            0.04  0.995833  0.691176   0.230392     0.996319  0.345588
4            0.05  0.994862  0.573529   0.169935     0.995534  0.262185
5            0.06  0.994160  0.213235   0.068884     0.995405  0.104129
6            0.07  0.994125  0.198529   0.064286     0.995393  0.097122
7            0.08  0.993879  0.220588   0.067114     0.995112  0.102916
8            0.09  0.993984  0.191176   0.060465     0.995264  0.091873
9            0.10  0.994019  0.227941   0.070938     0.995241  0.108202
10           0.11  0.994008  0.198529   0.062791     0.995276  0.095406
11           0.12  0.994019  0.198529   0.062937     0.995288  0.095575
12           0.13  0.993949  0.198529   0.062069     0.995217  0.094571
13           0.14  0.977740  0.264706   0.019587     0.978876  0.036474
14           0.15  0.974205  0.264706   0.016822     0.975336  0.031634
15           0.16  0.973725  0.264706   0.016506     0.974856  0.031075
16           0.17  0.973479  0.257353   0.015909     0.974621  0.029966
17           0.18  0.973749  0.257353   0.016077     0.974891  0.030264
18           0.19  0.973620  0.257353   0.015996     0.974762  0.030120
19           0.20  0.974252  0.257353   0.016401     0.975395  0.030837
20           0.21  0.973199  0.257353   0.015737     0.974340  0.029661
21           0.22  0.972765  0.257353   0.015480     0.973906  0.029203
22           0.23  0.972918  0.257353   0.015569     0.974058  0.029362
23           0.24  0.972847  0.272059   0.016386     0.973965  0.030911
24           0.25  0.973046  0.242647   0.014778     0.974211  0.027860
25           0.26  0.972344  0.242647   0.014392     0.973507  0.027172
26           0.27  0.971888  0.242647   0.014151     0.973050  0.026742
27           0.28  0.971923  0.242647   0.014169     0.973085  0.026775
28           0.29  0.971221  0.250000   0.014220     0.972370  0.026909

* **C = Contamination**: Cuántos datos sospechosos descartamos.
* **Samples Used:** Cuántos datos quedan después de descartar los outliers.
* **Accuracy (Exactitud):** Cuántas predicciones hicimos bien en total.
* **Recall (Sensibilidad):** De todos los fraudes reales, ¿cuántos detectamos?
* **Precision (Precisión):** De todas las transacciones que dijimos que eran fraudes, ¿cuántas realmente eran fraudes?
* **Specificity:** De todas las transacciones normales, ¿cuántas detectamos correctamente?
* **F1-score:** Un balance entre precisión y recall (si es alto, el modelo es bueno detectando fraudes).



In [ ]:
#@title 3. Encontrar el mejor F1-score
best_row = metrics_df.loc[metrics_df["F1-score"].idxmax()]
best_c = best_row["Contaminación"] #mejor contaminación

print(f"Mejor contaminación para Determinante de Covarianza: {best_c:.2f}")
#print(best_row)  # Muestra la información de la fila con el mejor F1-SCORE

Mejor contaminación para Determinante de Covarianza: 0.01


In [ ]:
#@title 4. Entrenar nuevamente con la mejor contaminación seleccionada
best_mcd = EllipticEnvelope(contamination=best_c, random_state=42)
best_mcd.fit(X_train_scaled)

y_pred_best = best_mcd.predict(X_train_scaled)
mask_best = y_pred_best != -1  # Filtrar datos sin outliers
X_train_best, y_train_best = X_train_scaled[mask_best], y_train[mask_best]

# Entrenar el mejor modelo de Regresión Logística
best_model_mcd = LogisticRegression()
best_model_mcd.fit(X_train_best, y_train_best)

# Predecir en el conjunto de prueba
y_pred_best_mcd = best_model_mcd.predict(X_test_scaled)

/usr/local/lib/python3.11/dist-packages/sklearn/covariance/_robust_covariance.py:185: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-191.218180695156036 > -298.257078914440854). You may want to try with a higher value of support_fraction (current value: 0.501).
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/covariance/_robust_covariance.py:185: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-253.553537586707506 > -254.462736667838271). You may want to try with a higher value of support_fraction (current value: 0.501).
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/covariance/_robust_covariance.py:185: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-248.238403078166982 > -248.800545362113724). You may want to try with a higher value of support_fraction (current value: 0.501).
  warnings.warn(
/usr/local

In [ ]:
#@title 5. Matriz de confusión del mejor modelo
best_conf_matrix = confusion_matrix(y_test, y_pred_best_mcd)
best_conf_matrix

array([[85280,    27],
       [   26,   110]])

In [ ]:
#@title 6.Métricas de Evaluación
print("Accuracy: ", accuracy_score(y_test, y_pred_best_mcd))
print("Recall: ", recall_score(y_test, y_pred_best_mcd))
print("precision_score: ", precision_score(y_test, y_pred_best_mcd))
print("Especificidad: ", recall_score(y_test, y_pred_best_mcd, pos_label=0)) # Negativos predichos correctamente
print("F1-score: ", f1_score(y_test, y_pred_best_mcd))

Accuracy:  0.9993797034280163
Recall:  0.8088235294117647
precision_score:  0.8029197080291971
Especificidad:  0.9996834960788681
F1-score:  0.8058608058608059


**ASPECTO POSITIVO:**

* **Recall mejoró:** (de 62.5% a 80.88%), lo que significa que el **modelo detecta muchos más fraudes.**
* **F1-score aumentó** de 72.9% a 80.58%, lo que indica un mejor balance entre precisión y recall.

**ASPECTO NEGATIVO:**
* **La precision bajó** ligeramente (87.6% → 80.3%), lo que significa que el modelo está detectando más fraudes, pero con más falsos positivos.
* **La especificidad bajó un poco**, pero sigue siendo casi perfecta (~99.97%).
<br><br>

**SIGUIENTE PASO:** Probar Isolation Forest para ver si mejora el recall aún más.

## **2. Pregunta**
Utiliza el Isolation Forest para detectar los outliers y hacer predicciones con el mismo. Calcula las métricas y muestra la matriz de confusión


---



In [ ]:
#@title 1.Isolation Forest y Entrenar
from sklearn.ensemble import IsolationForest
best_iforest = IsolationForest(contamination=0.01, random_state=42)
best_iforest.fit(X_train_scaled)

y_pred_best_iforest = best_iforest.predict(X_train_scaled)
mask_best_iforest = y_pred_best_iforest == 1  # Filtrar datos sin outliers
X_train_best_iforest, y_train_best_iforest = X_train_scaled[mask_best_iforest], y_train[mask_best_iforest]

# Entrenar modelo de Regresión Logística
best_model_iforest = LogisticRegression()
best_model_iforest.fit(X_train_best_iforest, y_train_best_iforest)

# Predecir en el conjunto de prueba
y_pred_best_iforest_test = best_model_iforest.predict(X_test_scaled)

In [ ]:
#@title 2. Matriz de confusión del mejor modelo Isolation Forest
best_conf_matrix_iforest = confusion_matrix(y_test, y_pred_best_iforest_test)
print("Matriz de Confusión del Mejor Modelo Isolation Forest:")
print(best_conf_matrix_iforest)


Matriz de Confusión del Mejor Modelo Isolation Forest:
[[85265    42]
 [   26   110]]


In [ ]:
#@title 3.Métricas de Evaluación
print("Accuracy: ", accuracy_score(y_test, y_pred_best_iforest_test))
print("Recall: ", recall_score(y_test, y_pred_best_iforest_test))
print("precision_score: ", precision_score(y_test, y_pred_best_iforest_test))
print("Especificidad: ", recall_score(y_test, y_pred_best_iforest_test, pos_label=0)) # Negativos predichos correctamente
print("F1-score: ", f1_score(y_test, y_pred_best_iforest_test))

Accuracy:  0.999204147794436
Recall:  0.8088235294117647
precision_score:  0.7236842105263158
Especificidad:  0.9995076605671281
F1-score:  0.7638888888888888


**ASPECTOS POSITIVOS:**

* El **Recall se mantiene (80.88%)**, igual que con MCD.
* Se detectan más fraudes, pero se generan más falsos positivos (baja precisión).


**ASPECTOS NEGATIVOS:**
* **La precisión bajó más** que con MCD: 80.3% → 72.3% . Ahora casi el 30% de los fraudes detectados son errores.
* El **F1-score bajó** un poco (80.5% → 76.3%).


---

<br>

* **Isolation Forest es útil, pero genera demasiados falsos positivos.**
* **El modelo con MCD sigue siendo mejor** (balance entre Recall y Precisión).


## **3. Pregunta**
3. ¿Qué porcentaje de predicciones como positivas eran correctas?
4. ¿Qué porcentaje de casos que eran positivos se han predicho correctamente?
5. ¿Qué porcentaje de casos que eran negativos se han predicho correctamente?


---



In [ ]:
# Pregunta 3: ¿Qué porcentaje de predicciones como positivas eran correctas? (Precisión)
precision = precision_score(y_test, y_pred_best_iforest_test)
print(f"Precisión (Porcentaje de fraudes predichos correctamente): {precision:.4f}")

# Pregunta 4: ¿Qué porcentaje de casos que eran positivos se han predicho correctamente? (Recall)
recall = recall_score(y_test, y_pred_best_iforest_test)
print(f"Recall (Porcentaje de fraudes detectados correctamente): {recall:.4f}")

# Pregunta 5: ¿Qué porcentaje de casos que eran negativos se han predicho correctamente? (Especificidad)
specificity = recall_score(y_test, y_pred_best_iforest_test, pos_label=0)
print(f"Specificity (Porcentaje de transacciones normales detectadas correctamente): {specificity:.4f}")


Precisión (Porcentaje de fraudes predichos correctamente): 0.7237
Recall (Porcentaje de fraudes detectados correctamente): 0.8088
Specificity (Porcentaje de transacciones normales detectadas correctamente): 0.9995


## **4. Pregunta**

6. Utiliza Logistic Regression sensible al coste y comprueba las métricas. ¿Ha mejorado el Recall? Imprime la matriz de confusión
7. Haz lo mismo haciendo undersampling
8. Haz lo mismo con oversampling
9. Haz lo mismo con SMOTE


---



In [ ]:
#@title 6.Cost-Sensitive Training
from sklearn.linear_model import LogisticRegression

# Entrenar el modelo de Regresión Logística con penalización por desbalance
cost_sensitive_model = LogisticRegression(class_weight="balanced", random_state=42)
cost_sensitive_model.fit(X_train_scaled, y_train)

# Predecir en el conjunto de prueba
y_pred_cost_sensitive = cost_sensitive_model.predict(X_test_scaled)

# Evaluar el modelo
print("Accuracy:", accuracy_score(y_test, y_pred_cost_sensitive))
print("Recall:", recall_score(y_test, y_pred_cost_sensitive))
print("Precision:", precision_score(y_test, y_pred_cost_sensitive))
print("Specificity:", recall_score(y_test, y_pred_cost_sensitive, pos_label=0))
print("F1-score:", f1_score(y_test, y_pred_cost_sensitive))

# Matriz de confusión
conf_matrix_cost_sensitive = confusion_matrix(y_test, y_pred_cost_sensitive)
print("\nMatriz de Confusión - Cost-Sensitive Training:")
print(conf_matrix_cost_sensitive)


Accuracy: 0.97424013669932
Recall: 0.9264705882352942
Precision: 0.054380664652567974
Specificity: 0.9743162929185178
F1-score: 0.10273134936812067

Matriz de Confusión - Cost-Sensitive Training:
[[83116  2191]
 [   10   126]]




---


Detecta muchos fraudes (**alto Recall**), pero genera **demasiados falsos positivos.**
Baja precisión significa que muchas transacciones legítimas son clasificadas como fraude.


In [ ]:
#@title  7.Undersampling:
from imblearn.under_sampling import RandomUnderSampler

# Aplicar Undersampling
undersampler = RandomUnderSampler(random_state=42)
X_train_under, y_train_under = undersampler.fit_resample(X_train_scaled, y_train)

# Entrenar modelo
model_under = LogisticRegression()
model_under.fit(X_train_under, y_train_under)

# Predecir
y_pred_under = model_under.predict(X_test_scaled)

# Evaluar el modelo
print("Accuracy:", accuracy_score(y_test, y_pred_under))
print("Recall:", recall_score(y_test, y_pred_under))
print("Precision:", precision_score(y_test, y_pred_under))
print("Specificity:", recall_score(y_test, y_pred_under, pos_label=0))
print("F1-score:", f1_score(y_test, y_pred_under))

# Evaluar
print("\nMatriz de Confusión - Undersampling:")
print(confusion_matrix(y_test, y_pred_under))


Accuracy: 0.9607925751670704
Recall: 0.9338235294117647
Precision: 0.036620530565167245
Specificity: 0.9608355703517882
F1-score: 0.07047724750277469

Matriz de Confusión - Undersampling:
[[81966  3341]
 [    9   127]]




---
**Aumenta el Recall a 93.3%**, pero con una **precisión muy baj**a.
Muchos falsos positivos, lo que no es ideal en un sistema de detección de fraudes.



In [ ]:
#@title  8.Oversampling:
from imblearn.over_sampling import RandomOverSampler

# Aplicar Oversampling
oversampler = RandomOverSampler(random_state=42)
X_train_over, y_train_over = oversampler.fit_resample(X_train_scaled, y_train)

# Entrenar modelo
model_over = LogisticRegression()
model_over.fit(X_train_over, y_train_over)

# Predecir
y_pred_over = model_over.predict(X_test_scaled)

# Evaluar el modelo
print("Accuracy:", accuracy_score(y_test,y_pred_over))
print("Recall:", recall_score(y_test, y_pred_over))
print("Precision:", precision_score(y_test, y_pred_over))
print("Specificity:", recall_score(y_test, y_pred_over, pos_label=0))
print("F1-score:", f1_score(y_test, y_pred_over))


# Evaluar
print("\nMatriz de Confusión - Oversampling:")
print(confusion_matrix(y_test, y_pred_over))


Accuracy: 0.974006062521213
Recall: 0.9264705882352942
Precision: 0.05391527599486521
Specificity: 0.9740818455695313
F1-score: 0.101900525677315

Matriz de Confusión - Oversampling:
[[83096  2211]
 [   10   126]]




---


* Resultados similares a Cost-Sensitive Training.
* Se detectan muchos fraudes, pero con muchos falsos positivos.

In [ ]:
#@title 9.SMOTE:

from imblearn.over_sampling import SMOTE

# Aplicar SMOTE
smote = SMOTE(random_state=42)
X_train_smote, y_train_smote = smote.fit_resample(X_train_scaled, y_train)

# Entrenar modelo
model_smote = LogisticRegression()
model_smote.fit(X_train_smote, y_train_smote)

# Predecir
y_pred_smote = model_smote.predict(X_test_scaled)

# Evaluar el modelo
print("Accuracy:", accuracy_score(y_test,y_pred_smote))
print("Recall:", recall_score(y_test, y_pred_smote))
print("Precision:", precision_score(y_test, y_pred_smote))
print("Specificity:", recall_score(y_test, y_pred_smote, pos_label=0))
print("F1-score:", f1_score(y_test, y_pred_smote))

# Evaluar
print("\nMatriz de Confusión - SMOTE:")
print(confusion_matrix(y_test, y_pred_smote))




Accuracy: 0.9725899137436653
Recall: 0.9264705882352942
Precision: 0.05126118795768918
Specificity: 0.9726634391081623
F1-score: 0.09714726291441789

Matriz de Confusión - SMOTE:
[[82975  2332]
 [   10   126]]




---


* Mejor que Undersampling, pero peor que Oversampling en términos de precisión.
* Aumenta la detección de fraudes, pero con muchos falsos positivos.


**CONCLUSIÓN**

* **MCD** fue la mejor técnica con **Recall del 80.88%** y un buen equilibrio entre Recall y Precisión.
* **Isolation Forest** también fue útil, pero con **más falsos positivos**.